src/ptychography40/reconstruction/wdd/wdd_udf.py

```diff
@@ -204,7 +223,275 @@ def get_results(self):
+    method is experimental and not fully validated.
+
+    Parameters
+    ----------
+    recon_parameters
+        Reconstruction plan generated by :func:`ptychography40.reconstruction.wdd.\
+            params_recon.wdd_params_recon`
+    patch_overlap
+        (y, x) overlap of reconstruction patches after subtracting trim, default is two
+    patch_trim
+        Remove border pixels of patches to suppress edge effects, default 2 pixel
+    complex_dtype
+        Complex dtype for array
+
+    """
+
+    def __init__(self, recon_parameters: WDDReconstructionPlan,
+                 patch_overlap: Optional[Tuple] = None,
+                 patch_trim: Optional[Union[Tuple, int]] = None,
+                 complex_dtype: "nt.DTypeLike" = np.complex64):
+        patch_y, patch_x = recon_parameters.patch_shape
+        if patch_trim is None:
+            patch_trim = (8, 8)
+        elif isinstance(patch_trim, int):
+            patch_trim = (patch_trim, patch_trim)
+        if patch_overlap is None:
+            patch_overlap = (8, 8)
+        elif isinstance(patch_overlap, int):
+            patch_overlap = (patch_overlap, patch_overlap)
+
+        patch_step = self._calc_patch_step(
+            patch_shape=(patch_y, patch_x),
+            patch_trim=patch_trim,
+            patch_overlap=patch_overlap
+        )
+
+        if patch_step[0] <= 0 or patch_step[1] <= 0:
+            raise ValueError(
+                f"Patch step {patch_step} <= 0, patch overlap {patch_overlap} and"
+                F"patch trim {patch_trim} are too large for patch size ({patch_y, patch_x}).")
+        super().__init__(
+            recon_parameters=recon_parameters,
+            patch_overlap=patch_overlap,
+            patch_trim=patch_trim,
+            complex_dtype=complex_dtype
+        )
+
+    @staticmethod
+    def _calc_patch_step(patch_shape, patch_trim, patch_overlap):
+        return (
+            patch_shape[0] - 2*patch_trim[0] - patch_overlap[0],
+            patch_shape[1] - 2*patch_trim[1] - patch_overlap[1]
+        )
+
+    @property
+    def patch_step(self):
+        return self._calc_patch_step(
+            patch_shape=self.patch_shape,
+            patch_trim=self.params.patch_trim,
+            patch_overlap=self.params.patch_overlap
+        )
+
+    @property
+    def patch_shape(self):
+        'Shorthand'
+        return self.params.recon_parameters.patch_shape
+
+    @property
+    def gridspec(self):
+        patch_shape = self.patch_shape
+        patch_step = self.patch_step
+        y_grid = GridSpec(step=patch_step[0], size=patch_shape[0])
+        x_grid = GridSpec(step=patch_step[1], size=patch_shape[1])
+        return (y_grid, x_grid)
+
+    @property
+    def patch_offset(self):
+        (y_grid, x_grid) = self.gridspec
+        return (-y_grid.min_patch_index, -x_grid.min_patch_index)
+
+    def get_result_buffers(self) -> dict[str, BufferWrapper]:
```

Comment:
`python_requires` of this repo is still 3.7+, so we need to either use `typing.Dict` or bump `python_requires` accordingly